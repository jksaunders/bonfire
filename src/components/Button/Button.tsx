import React, { FC } from 'react';
import styled from 'styled-components';

const StyledButton = styled.button`
  background-color: ${(props): string => props.color};
`;

interface ButtonProps {
  color: string;
  children?: never;
}

const Button: FC<ButtonProps> = ({ color = 'red' }) => (
  <StyledButton color={color}>test button</StyledButton>
);

export default Button;
