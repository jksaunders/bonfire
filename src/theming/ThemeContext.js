import React, { useState, useContext } from 'react';
import { isArray } from 'util';

const ThemeContext = React.createContext(null);

const useTheme = (
  initialTheme = {
    current: {},
    variants: {
      components: { button: {}, typography: {} },
    },
  }
) => {
  const [theme, setTheme] = useState(initialTheme);

  const setCurrentTheme = (newCurrent) => {
    setTheme({
      ...theme,
      current: {
        ...theme.current,
        ...newCurrent,
      },
    });
  };

  const getCurrentVariant = (componentKey, variantKey) => {
    const result = {};

    if (theme.variants == null) {
      throw new Error('Theme missing `variants` field');
    }
    if (theme.variants.components == null) {
      throw new Error('`theme.variants` missing `components` field');
    }
    if (theme.variants.components[componentKey] == null) {
      throw new Error(
        `\`theme.variants.components\` missing \`${componentKey}\` field`
      );
    }
    if (theme.variants.components[componentKey][variantKey] == null) {
      throw new Error(
        `\`theme.variants.components.${componentKey}\` missing \`${variantKey}\` field`
      );
    }

    const variant = theme.variants.components[componentKey][variantKey];

    if (typeof variant !== 'object') {
      return {};
    }

    Object.keys(variant).forEach((propKey) => {
      const variantValue =
        theme.variants.components[componentKey][variantKey][propKey];

      result[propKey] =
        typeof variantValue === 'function'
          ? variantValue(theme.current)
          : variantValue;
    });

    return result;
  };

  return {
    theme,
    getCurrentVariant,
    setCurrentTheme,
  };
};

export const bonfireThemePropKey = 'bonfireTheme';
export const bonfireThemeVariantPropKey = 'variant';

export const getVariant = (
  componentName,
  props,
  defaultProps,
  propsToCheck
) => {
  const result = {};

  const variant = props[bonfireThemeVariantPropKey];
  const themeVariant =
    typeof variant === 'string' && props[bonfireThemePropKey] != null
      ? props[bonfireThemePropKey].getCurrentVariant(componentName, variant)
      : null;

  const toCheck = isArray(propsToCheck)
    ? propsToCheck
    : Object.keys(propsToCheck);

  const objectCheck = (o) => o != null && typeof o === 'object' && !isArray(o);

  toCheck.forEach((k) => {
    if (
      objectCheck(props) &&
      objectCheck(props[k]) &&
      objectCheck(themeVariant) &&
      objectCheck(themeVariant[k])
    ) {
      result[k] = {
        ...themeVariant[k],
        ...props[k],
      };
    } else if (props[k] != null) {
      result[k] = props[k];
    } else if (themeVariant != null && themeVariant[k] != null) {
      result[k] = themeVariant[k];
    } else if (objectCheck(variant) && variant[k] != null) {
      result[k] = variant[k];
    } else if (defaultProps[k] != null) {
      result[k] = defaultProps[k];
    }
  });
  return result;
};

export const withThemes = (Component) =>
  React.forwardRef((props, ref) => {
    const theme = useContext(ThemeContext);
    const bonfireProps = { [bonfireThemePropKey]: theme };
    return <Component {...props} {...bonfireProps} ref={ref} />;
  });

// eslint-disable-next-line react/prop-types
export const ThemeRoot = ({ children, initialTheme }) => {
  const themeRoot = useTheme(initialTheme);
  return (
    <ThemeContext.Provider value={themeRoot}>{children}</ThemeContext.Provider>
  );
};

export default ThemeContext;
