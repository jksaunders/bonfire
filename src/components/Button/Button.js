import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import Typography, { typographyPropTypes } from '../Typography';
import * as Colors from '../../theming/colors';
import { getVariant, withThemes } from '../../theming/ThemeContext';
import { css, cssBackground } from '../../styling';

const StyledButton = styled.button`
  ${props => cssBackground(props, { alpha: props.disabled ? 0.5 : 1 })}
  ${css('borderRadius', 'border-radius')}
  border-style: solid;
  border-width: 1px;
  ${css('padding')}
  width: ${({ width }) => width};
`;

const buttonStyleProps = {
  background: PropTypes.string,
  borderRadius: PropTypes.string,
  borderStyle: PropTypes.string,
  borderWidth: PropTypes.string,
  padding: PropTypes.string,
  typography: PropTypes.shape(typographyPropTypes),
};

const propTypes = {
  ...buttonStyleProps,
  disabled: PropTypes.bool,
  onClick: PropTypes.func,
  text: PropTypes.string.isRequired,
  variant: PropTypes.oneOfType([
    PropTypes.shape(buttonStyleProps),
    PropTypes.string,
  ]),
  width: PropTypes.string,
};

const defaultProps = {
  /* eslint-disable react/default-props-match-prop-types */
  background: Colors.blue,
  borderRadius: null,
  borderStyle: null,
  borderWidth: null,
  padding: null,
  typography: null,
  /* eslint-enable react/default-props-match-prop-types */
  disabled: false,
  onClick: () => {},
  variant: null,
  width: '100px',
};

const Button = ({ disabled, onClick, text, ...props }) => {
  const { typography, ...buttonVariant } = getVariant(
    'button',
    props,
    buttonStyleProps
  );
  return (
    <StyledButton
      disabled={disabled}
      onClick={onClick}
      text={text}
      {...buttonVariant}
    >
      <Typography {...typography}>{text}</Typography>
    </StyledButton>
  );
};

Button.propTypes = propTypes;
Button.defaultProps = defaultProps;

const ThemedButton = withThemes(Button);

export default ThemedButton;
