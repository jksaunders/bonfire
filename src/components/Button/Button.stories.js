import React from "react";
import { ThemeProvider } from "styled-components";

import { action } from "@storybook/addon-actions";
import { select, withKnobs } from "@storybook/addon-knobs";

import { ThemeConstants } from "../../theming";
import Button from "./Button";

export default {
  title: "Components|Button",
  component: Button,
  decorators: [withKnobs]
};

export const ButtonStory = () => {
  const themeMode = select("Theme Mode", Object.keys(ThemeConstants.mode.values), Object.keys(ThemeConstants.mode.values)[0]);
  const themeLayout = select("Theme Layout", Object.keys(ThemeConstants.layout.values), Object.keys(ThemeConstants.layout.values)[0]);
  const theme = { mode: themeMode, layout: themeLayout };

  return (
    <ThemeProvider theme={theme}>
      <div>
        <div>
          {Object.keys(Button.CONSTANTS.VARIANT).map(variantKey => (
            <Button key={variantKey} text={variantKey.toLowerCase()} variant={variantKey} onClick={action("clicked")} />
          ))}
        </div>
        <div>
          {Object.keys(Button.CONSTANTS.VARIANT).map(variantKey => (
            <Button key={variantKey} text={`${variantKey.toLowerCase()} disabled`} variant={variantKey} onClick={action("clicked")} />
          ))}
        </div>
      </div>
    </ThemeProvider>
  );
};

ButtonStory.story = {
  name: "Button"
};