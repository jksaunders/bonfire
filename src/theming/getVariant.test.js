import React from 'react';
import { expectExists, expectSnapshot, expectStyle, render } from '../testing';
import Button from '../components/Button';
import { ThemeRoot } from './ThemeContext';

const initialTheme = {
  current: {},
  variants: {
    components: {
      button: {
        primary: {
          padding: '15px 20px',
          typography: {
            color: 'red',
            size: '14px',
          },
        },
      },
    },
  },
};

describe('variant and prop both present', () => {
  describe('string value', () => {
    test('using same keys: props take priority', () => {
      const { component, getByText } = render(
        <ThemeRoot initialTheme={initialTheme}>
          <Button text="content" padding="5px 5px" variant="primary" />
        </ThemeRoot>
      );
      expectExists(component);
      expectSnapshot(component);
      expectStyle(component, 'padding', '5px 5px');

      const span = getByText('content');
      expectExists(span);
      expectStyle(span, 'color', 'red');
      expectStyle(span, 'font-size', '14px');
    });
  });

  describe('object value', () => {
    test('using different keys: object merged', () => {
      const { component, getByText } = render(
        <ThemeRoot initialTheme={initialTheme}>
          <Button
            text="content"
            typography={{ align: 'left' }}
            variant="primary"
          />
        </ThemeRoot>
      );
      expectExists(component);
      expectSnapshot(component);
      expectStyle(component, 'padding', '15px 20px');

      const span = getByText('content');
      expectExists(span);
      expectStyle(span, 'color', 'red');
      expectStyle(span, 'font-size', '14px');
      expectStyle(span, 'text-align', 'left');
    });

    test('using same keys: props take priority', () => {
      const { component, getByText } = render(
        <ThemeRoot initialTheme={initialTheme}>
          <Button
            text="content"
            typography={{ size: '5px' }}
            variant="primary"
          />
        </ThemeRoot>
      );
      expectExists(component);
      expectSnapshot(component);
      expectStyle(component, 'padding', '15px 20px');

      const span = getByText('content');
      expectExists(span);
      expectStyle(span, 'color', 'red');
      expectStyle(span, 'font-size', '5px');
    });
  });
});
