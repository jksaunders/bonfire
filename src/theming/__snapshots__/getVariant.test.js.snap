// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`variant and prop both present object value using different keys: object merged 1`] = `
.c0 {
  box-sizing: border-box;
  border: none;
  cursor: pointer;
  padding: 15px 20px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  text-align: left;
  color: red;
  font-size: 14px;
  white-space: nowrap;
}

.c1 {
  cursor: pointer;
}

<button
  class="c0 c1"
  cursor="pointer"
>
  <span
    class="c2"
    color="red"
  >
    content
  </span>
</button>
`;

exports[`variant and prop both present object value using same keys: props take priority 1`] = `
.c0 {
  box-sizing: border-box;
  border: none;
  cursor: pointer;
  padding: 15px 20px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  color: red;
  font-size: 5px;
  white-space: nowrap;
}

.c1 {
  cursor: pointer;
}

<button
  class="c0 c1"
  cursor="pointer"
>
  <span
    class="c2"
    color="red"
  >
    content
  </span>
</button>
`;

exports[`variant and prop both present string value using same keys: props take priority 1`] = `
.c0 {
  box-sizing: border-box;
  border: none;
  cursor: pointer;
  padding: 5px 5px;
}

.c2 {
  margin: 0;
  -webkit-font-smoothing: antialiased;
  color: red;
  font-size: 14px;
  white-space: nowrap;
}

.c1 {
  cursor: pointer;
}

<button
  class="c0 c1"
  cursor="pointer"
>
  <span
    class="c2"
    color="red"
  >
    content
  </span>
</button>
`;
